lib/dns-resolver.js

```diff
@@ -0,0 +1,516 @@
+      }, fallbackDuration * 1000)
+
+      /* istanbul ignore next: There is no `interval.unref()` when running inside an Electron renderer */
+      if (interval.unref) {
+        interval.unref()
+      }
+    }
+
+    if (lookupOptions) {
+      this.lookup = (hostname, _options, callback) => this.#lookup(hostname, lookupOptions, callback)
+    } else {
+      this.lookup = this.#lookup.bind(this)
+    }
+    this.lookupAsync = this.lookupAsync.bind(this)
+  }
+
+  get [kDnsCacheSize] () {
+    return this.#cache.size ?? 0
+  }
+
+  get [kDnsHostnamesToFallback] () {
+    return this.#hostnamesToFallback.size ?? 0
+  }
+
+  set servers (servers) {
+    this.clear()
+
+    this.#resolver.setServers(servers)
+  }
+
+  get servers () {
+    return this.#resolver.getServers()
+  }
+
+  #lookup (hostname, options, callback) {
+    if (typeof options === 'function') {
+      callback = options
+      options = {}
+    } else if (typeof options === 'number') {
+      options = {
+        family: options
+      }
+    }
+
+    if (!callback) {
+      throw new Error('Callback must be a function.')
+    }
+
+    // eslint-disable-next-line promise/prefer-await-to-then
+    this.lookupAsync(hostname, options).then((result) => {
+      if (options.all) {
+        callback(null, result)
+      } else {
+        callback(
+          null,
+          result.address,
+          result.family,
+          result.expires,
+          result.ttl
+        )
+      }
+    }, callback)
+  }
+
+  async lookupAsync (hostname, options = {}) {
+    if (typeof options === 'number') {
+      options = {
+        family: options
+      }
+    }
+
+    let cached = await this.query(hostname)
+
+    if (options.family === 6) {
+      const filtered = []
+      for (const entry of cached) {
+        if (entry.family === 6) {
+          filtered.push(entry)
+        }
+      }
+      if (options.hints & V4MAPPED) {
+        if ((options.hints & ALL) || filtered.length === 0) {
+          map4to6(cached)
+        } else {
+          cached = filtered
+        }
+      } else {
+        cached = filtered
+      }
+    } else if (options.family === 4) {
+      const filtered = []
+      for (const entry of cached) {
+        if (entry.family === 4) {
+          filtered.push(entry)
+        }
+      }
+      cached = filtered
+    }
+
+    if (options.hints & ADDRCONFIG) {
+      const filtered = []
+      for (const entry of cached) {
+        if (entry.family === 6 && this.#iface.has6) {
+          filtered.push(entry)
+        } else if (entry.family === 4 && this.#iface.has4) {
+          filtered.push(entry)
+        }
+      }
+      cached = filtered
+    }
+
+    if (cached.length === 0) {
+      const error = new Error(`DNSResolver ENOTFOUND ${hostname}`)
+      error.code = 'ENOTFOUND'
+      error.hostname = hostname
+
+      throw error
+    }
+
+    if (options.all) {
+      return cached
+    }
+
+    if (this.#scheduling === 'first') {
```

Comment:
Wondering, what are the intentions with the `scheduling` option?
Its not added within `cacheable-lookup` as it already caches the records

Reply:
So with `first` it is the same behavior as `cacheable-lookup` and returns the first cached entry from cache.

With `random` it will return a random item from the cache.

This is a way to achieve some sort of round robin without to much complexity.